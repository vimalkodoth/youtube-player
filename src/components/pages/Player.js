import React, { useRef } from "react";
import { css } from "@emotion/core";
import YouTube from 'react-youtube';
import SubmitButton from "../SubmitButton";
import usePlaybackConfiguration from "../hooks/usePlaybackConfiguration";
import useStorePositionTime from "../hooks/useStorePositionTime";
import { useHistory } from "react-router-dom";

const defaultConfigs = {
    height: '390',
    width: '640',
    playerVars: {
        // https://developers.google.com/youtube/player_parameters
        autoplay: 1,
        start: 0
    }
};

function Player() {
    const playerRef = useRef(null);
    const opts = usePlaybackConfiguration(defaultConfigs);
    const [onPlayerStateChange] = useStorePositionTime(playerRef);
    const history = useHistory();
    const onEdit = () => {
        history.goBack();
    }
    if (!opts) return <div></div>;
    return (
        <div css={Wrapper}>
            <YouTube videoId="RrTlbxFhK_I" opts={opts} ref={playerRef} onStateChange={onPlayerStateChange} />
            <SubmitButton onClick={onEdit}>
                Edit
            </SubmitButton>
            <SubmitButton to={{
                pathname: `/gif`,
                state: {}
            }}>
                Next (GIF)
            </SubmitButton>
        </div>
    );
}

export default Player;

const Wrapper = css`
    display: flex;
    flex-direction: column;
    height: 100vh;
    justify-content: center;
    align-items: center;
    border: 1px solid #fff;
`;