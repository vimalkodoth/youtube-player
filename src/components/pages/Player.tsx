import React, { useEffect, useRef, useState } from "react";
import { css } from "@emotion/core";
import YouTube, { Options, YouTubeProps } from 'react-youtube';
import SubmitButton from "../SubmitButton";
import usePlaybackConfiguration from "../hooks/usePlaybackConfiguration";
import useStorePositionTime from "../hooks/useStorePositionTime";
import { useHistory } from "react-router-dom";
import useParseStreamUrl from "../hooks/useParseStreamUrl";
import { setPlayerState } from "../../actionCreators";
import { useDispatch } from 'react-redux';

const defaultConfigs: Options = {
    height: '390',
    width: '640',
    playerVars: {
        // https://developers.google.com/youtube/player_parameters
        autoplay: 1,
        start: 0
    }
};

interface IProps {
    streamUrl: string,
    location: any
}

const Player: React.FC<IProps> = ({ streamUrl, location }: IProps) => {
    const playerRef = useRef<YouTube | null>(null);
    const dispatch = useDispatch();
    const opts = usePlaybackConfiguration(defaultConfigs);
    const [onPlayerStateChange] = useStorePositionTime(playerRef, 2);
    const [streamId, error]: (string | boolean)[] = useParseStreamUrl(location.state.url);
    const history = useHistory();
    const onEdit = () => {
        dispatch(setPlayerState({
            seekPosition: -1,
            currentTime: -1
        }));
        history.goBack();
    }
    if (streamId === '' && error) return <div css={Wrapper}>An error has occured</div>;
    if (!opts || streamId === '') return <div css={Wrapper}></div>;
    return (
        <div css={Wrapper}>
            <YouTube videoId={streamId as string} opts={opts as Options} ref={playerRef} onStateChange={onPlayerStateChange} />
            <SubmitButton onClick={onEdit}>
                Edit
            </SubmitButton>
            <SubmitButton to={{
                pathname: `/gif`,
                state: {}
            }}>
                Next (GIF)
            </SubmitButton>
        </div>
    );
}

export default Player;

const Wrapper = css`
    display: flex;
    flex-direction: column;
    height: 100vh;
    justify-content: center;
    align-items: center;
    border: 1px solid #fff;
`;