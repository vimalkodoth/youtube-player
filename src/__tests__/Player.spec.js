import React from "react";
import { mount } from "enzyme";
import { Provider } from "react-redux";
import { MemoryRouter } from "react-router-dom";
import configureStore from "redux-mock-store";
import Player from "../components/pages/Player";

let store;
const mockStore = configureStore({});
const originalConsoleError = console.error;

describe("Player", () => {
    beforeEach(() => {
        console.error = jest.fn();
        store = mockStore({
            playerState: {}
        });
    });

    afterEach(() => {
        console.error = originalConsoleError;
        store = mockStore({});
    });

    it("renders correctly", () => {
        const component = mount(
            <Provider store={store}>
                <MemoryRouter>
                    <Player />
                </MemoryRouter>
            </Provider>
        );
        expect(component).toMatchSnapshot();
    });
});
