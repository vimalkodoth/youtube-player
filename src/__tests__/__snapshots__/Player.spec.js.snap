// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Player renders correctly 1`] = `
<Provider
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <MemoryRouter>
    <Router
      history={
        Object {
          "action": "POP",
          "block": [Function],
          "canGo": [Function],
          "createHref": [Function],
          "entries": Array [
            Object {
              "hash": "",
              "key": "tinwt1",
              "pathname": "/",
              "search": "",
              "state": undefined,
            },
          ],
          "go": [Function],
          "goBack": [Function],
          "goForward": [Function],
          "index": 0,
          "length": 1,
          "listen": [Function],
          "location": Object {
            "hash": "",
            "key": "tinwt1",
            "pathname": "/",
            "search": "",
            "state": undefined,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <Player>
        <EmotionCssPropInternal
          __EMOTION_LABEL_PLEASE_DO_NOT_USE__="Player"
          __EMOTION_TYPE_PLEASE_DO_NOT_USE__="div"
          css={
            Object {
              "name": "1vl48lr-Wrapper",
              "styles": "display:flex;flex-direction:column;height:100vh;justify-content:center;align-items:center;border:1px solid #fff;;label:Wrapper;",
            }
          }
        >
          <div
            className="css-1vl48lr-Wrapper"
          >
            <YouTube
              className={null}
              containerClassName=""
              id={null}
              onEnd={[Function]}
              onError={[Function]}
              onPause={[Function]}
              onPlay={[Function]}
              onPlaybackQualityChange={[Function]}
              onPlaybackRateChange={[Function]}
              onReady={[Function]}
              onStateChange={[Function]}
              opts={
                Object {
                  "height": "390",
                  "playerVars": Object {
                    "autoplay": 1,
                    "start": 1629584745,
                  },
                  "width": "640",
                }
              }
              videoId="RrTlbxFhK_I"
            >
              <div
                className=""
              >
                <div
                  className={null}
                  id={null}
                />
              </div>
            </YouTube>
            <SubmitButton
              onClick={[Function]}
            >
              <EmotionCssPropInternal
                __EMOTION_LABEL_PLEASE_DO_NOT_USE__="SubmitButton"
                __EMOTION_TYPE_PLEASE_DO_NOT_USE__="input"
                css={
                  Object {
                    "name": "2q23bc-Input",
                    "styles": "margin:0.75rem;margin-left:0;;label:Input;",
                  }
                }
                onClick={[Function]}
                type="submit"
                value="Edit"
              >
                <input
                  className="css-2q23bc-Input"
                  onClick={[Function]}
                  type="submit"
                  value="Edit"
                />
              </EmotionCssPropInternal>
            </SubmitButton>
            <SubmitButton
              to={
                Object {
                  "pathname": "/gif",
                  "state": Object {},
                }
              }
            >
              <Link
                to={
                  Object {
                    "pathname": "/gif",
                    "state": Object {},
                  }
                }
              >
                <LinkAnchor
                  href="/gif"
                  navigate={[Function]}
                >
                  <a
                    href="/gif"
                    onClick={[Function]}
                  >
                    <EmotionCssPropInternal
                      __EMOTION_LABEL_PLEASE_DO_NOT_USE__="SubmitButton"
                      __EMOTION_TYPE_PLEASE_DO_NOT_USE__="input"
                      css={
                        Object {
                          "name": "2q23bc-Input",
                          "styles": "margin:0.75rem;margin-left:0;;label:Input;",
                        }
                      }
                      type="submit"
                      value="Next (GIF)"
                    >
                      <input
                        className="css-2q23bc-Input"
                        type="submit"
                        value="Next (GIF)"
                      />
                    </EmotionCssPropInternal>
                  </a>
                </LinkAnchor>
              </Link>
            </SubmitButton>
          </div>
        </EmotionCssPropInternal>
      </Player>
    </Router>
  </MemoryRouter>
</Provider>
`;
